unittests/savanna_proposer_policy_tests.cpp

```diff
@@ -0,0 +1,159 @@
+   A.set_producers(producers);   // set new producers and produce blocks until the switch is pending
+   auto end_num = A.head().block_num();
+
+   // under Savanna, a new policy becomes active on the first block of a 12 block round after:
+   // 1. finishing the current round
+   // 2. a full round
+   // ----------------------------------------------------------------------------------------
+   uint32_t expected_gap = 12 - (start_num % 12);   // 1. finishing the current round
+   expected_gap += 12;                              // 2. a full round
+   BOOST_REQUIRE_EQUAL(end_num, start_num + expected_gap);
+} FC_LOG_AND_RETHROW()
+
+
+// ---------------------------------------------------------------------------------------------------
+//     Verify that a proposer policy does not become active when finality has stalled
+// ---------------------------------------------------------------------------------------------------
+BOOST_FIXTURE_TEST_CASE(no_proposer_policy_change_without_finality, savanna_cluster::cluster_t) try {
+   auto& A=_nodes[0];
+
+   // split network { A, B } and { C, D }
+   // Regardless of how many blocks A produces, finality will not advance
+   // by more than one (1 QC in flight)
+   // -------------------------------------------------------------------
+   const std::vector<size_t> partition {2, 3};
+   set_partition(partition);
+   A.produce_block();                     // take care of the in-flight QC
+
+   const vector<account_name> producers { "p0"_n, "p1"_n };
+   auto orig_version = A.control->active_producers().version;
+   auto orig_lib_num = A.lib_number;
+
+   A.create_accounts(producers);
+   A.tester::set_producers(producers);    // push the action to update the producer schedule
+
+   // produce 24 more blocks. If finality was advancing, the new proposer policy would be active,
+   // but with a split network finality will stall, and the new proposer policy should *not* become active.
+   // -----------------------------------------------------------------------------------------------------
+   A.produce_blocks(24);
+   BOOST_REQUIRE_EQUAL(A.lib_number, orig_lib_num);
+   BOOST_REQUIRE_EQUAL(A.control->active_producers().version, orig_version);
+} FC_LOG_AND_RETHROW()
+
+// ---------------------------------------------------------------------------------------------------
+//     Verify that a proposer policy becomes active when finality has advanced enough to make it pending
+// ---------------------------------------------------------------------------------------------------
+BOOST_FIXTURE_TEST_CASE(pending_proposer_policy_becomes_active_without_finality, savanna_cluster::cluster_t) try {
+   auto& A=_nodes[0];
+
+   const vector<account_name> producers { "p0"_n, "p1"_n };
+   auto orig_version = A.control->active_producers().version;
+
+   while (A.head().block_num() % 12 >= 10)
+      A.produce_block();                  // make sure we are not on the last two blocks of a round
```

Comment:
Again, this needs to be corrected to use the slot instead of the block number. 

And I think there may be an off-by-one error again, but it seems like it would only matter in terms of not matching the comment. It does not appear like it would harm the actual test.

For example, if the head block's time slot, `slot1`, happens to satisfy `slot1 % 12 == 9`, then no additional block would be produced. But that means the next pending block after that head block, the one that includes the transaction that will propose a new proposer policy, would have a slot, `slot2`, which satisfies `slot2 % 12 == 10`. If I'm not mistaken, that would mean it would be the second to last block of the round, which is something the comment says it wants to prevent. But anyway, after producing 13 blocks, the new pending block would have a slot, `slot3`, which satisfies `slot3 % 12 == 11` but is on the next round. So it wouldn't yet have the new proposer policy as active. That would happen on the very next block in this case. But because it is a new round, the new proposer policy would be pending.

It would also unnecessarily prevent the proposer policy from being proposed on the first block of a round, but that doesn't really harm anything either.